prereise/gather/griddata/hifld/data_process/generators.py

```diff
@@ -84,25 +86,198 @@ def map_generator_to_bus_by_sub(generator, bus_groupby):
+        elif gen_type in const.reasonable_heat_rates_size_cutoffs:
+            if generator["Pmax"] < const.reasonable_heat_rates_size_cutoffs[gen_type]:
+                key = tuple((*gen_type, "small"))
+            else:
+                key = tuple((*gen_type, "large"))
+            bounds = const.reasonable_heat_rates_by_type_and_size[key]
+        else:
+            bounds = (0, float("inf"))
+        return pd.Series(bounds, index=("lower", "upper"))
+
+    heat_rate_bound_types = set(const.reasonable_heat_rates_size_cutoffs.keys()) | set(
+        const.reasonable_heat_rates_by_type.keys()
+    )
+    generator_types = generators[["Technology", "Prime Mover"]].apply(tuple, axis=1)
+    gens_with_coefficients = generators.loc[
+        generator_types.isin(heat_rate_bound_types) & ~generators["h0"].isna()
+    ]
+    # Simplified calculation for rise/run for quadratic curve evaluated at Pmin and Pmax
+    single_segment_slope = (
+        gens_with_coefficients["h2"]
+        * (gens_with_coefficients["Pmax"] + gens_with_coefficients["Pmin"])
+        + gens_with_coefficients["h1"]
+    )
+    heat_rate_bounds = gens_with_coefficients.apply(get_heat_rate_bounds, axis=1)
+    good_heat_rates = single_segment_slope.loc[
+        (heat_rate_bounds["lower"] <= single_segment_slope)
+        & (single_segment_slope <= heat_rate_bounds["upper"])
+    ]
+    generators.loc[
+        ~generators.index.isin(good_heat_rates.index), ["h0", "h1", "h2"]
+    ] = float("nan")
+
+
+def augment_missing_heat_rates(generators):
+    """Identify generators with NaN heat rate coefficients, replace with quadratic
+    coefficients from a linear regression against other generators of that type (for
+    fossil generators) or with assumed linear coefficients for non-fossil generators.
+
+    :param pandas.DataFrame generators: data frame of generators with fitted heat rates
+        (modified inplace).
+    """
+
+    def estimate_coefficients(generator, regressions):
+        """Estimate heat rate coefficients for a generator, by type and capacity.
+
+        :param pandas.Series generator: single generator attributes.
+        :param dict regressions: nested dictionary of fitted regressions:
+            first level is ('Technology', 'Prime Mover') attributes,
+            second level is {'h0', 'h1', 'h2'}, values are regressions against Pmax.
+        :return: (*pandas.Series*) -- heat rate parameters: (h0, h1, h2).
+        """
+        regression_key = tuple(generator[["Technology", "Prime Mover"]])
+        if regression_key not in regressions:
+            return pd.Series([float("nan")] * 3, index=["h0", "h1", "h2"])
+        type_regressions = regressions[regression_key]
+        estimates = pd.Series(
+            {
+                coef: (regression.intercept + regression.slope * generator["Pmax"])
+                for coef, regression in type_regressions.items()
+            },
+            index=output_index,
```

Comment:
Do we need this, given the `coef` is the same as `output_index`?

Reply:
You're right, removed.